axlearn/common/trainer.py

```diff
@@ -581,7 +582,12 @@ def run(
                 output = None
                 stop_trace_step = None
 
-                input_iterator = self.input.batches(self._input_iter)
+                input_iterator = LoggingIterator(
```

Comment:
Since it is already recorded in the Goodput, I wonder if we need a separate logging here. 

If I am reading it correctly, this LoggingIterator creates one timer thread every time `next` is called, which seems to be adding a lot of thread creation/cancellation overhead to the main process

Reply:
Assuming you mean this code
```
                    self._maybe_record_event(measurement.Event.START_DATA_LOADING)
                    try:
                        input_batch = next(input_iterator)
                        self._maybe_record_event(measurement.Event.END_DATA_LOADING)
```

If the next() somehow hangs and then the whole process aborts, would Goodput measurement help catch that?

My understanding is this iterator invokes next() every a few seconds depending on the training scale, so creating and cleaning a thread for every a few seconds should have only very minimal overhead (empirically one thread takes 10 to 100 usec to create).

However, if you still have concerns on overhead, we might create a reusable thread for timer purpose.